index.mjs

```diff
@@ -0,0 +1,203 @@
+  try {
+    const args = [...GLOBAL_OSMDBT_ARGS, ...commandArgs];
+    const commandPath = path.join(OSMDBT_BIN_PATH, command);
+    await $`${commandPath} ${args}`;
+  } catch (error) {
+    if (onError) {
+      await onError();
+    }
+    throw new ErrorWithExitCode(error.stderr, ExitCodes.OSMDBT_ERROR);
+  }
+};
+
+const putObjectWrapper = async (key, body) => {
+  const { bucketName, acl } = objectStorageConfig;
+  try {
+    logger.info(`putting key: ${key} into bucket: ${bucketName}`);
+    await s3Client.send(new PutObjectCommand({ Bucket: bucketName, Key: key, Body: body, ACL: acl }));
+  } catch (error) {
+    throw new ErrorWithExitCode(
+      `failed putting key: ${key} into bucket: ${bucketName} received the following error: ${error}`,
+      ExitCodes.PUT_OBJECT_ERROR
+    );
+  }
+};
+
+const changeLogFilesToUndone = async () => {
+  const { logDir } = osmdbtConfig;
+  const logFilesNames = await fs.readdir(logDir);
+  logFilesNames.forEach(async (logFileName) => {
+    if (!logFileName.endsWith(OSMDBT_DONE_LOG_PREFIX)) {
+      return;
+    }
+    const logFileNameUndone = logFileName.slice(0, logFileName.length - OSMDBT_DONE_LOG_PREFIX.length);
+    const currentPath = path.join(logDir, logFileName);
+    const newPath = path.join(logDir, logFileNameUndone);
+    await fs.rename(currentPath, newPath);
+  });
+};
+
+const processExitSafely = (exitCode) => {
+  logger.info(`exiting safely with exit code: ${exitCode}`);
+
+  if (s3Client !== undefined) {
+    s3Client.destroy();
+  }
+  process.exit(exitCode);
+};
+
+const rollback = async () => {
+  logger.info('something went wrong running rollback');
+  await putObjectWrapper(STATE_FILE, fs.createReadStream(OSMDBT_STATE_BACKUP_PATH));
+}
+
+logger.info(`new job has started`);
+try {
+  await prepareEnvironment();
+
+  s3Client = initializeS3Client();
+  await getStateFileFromS3ToFs();
+  const lastSequenceNumber = await getSequenceNumber();
+
+  logger.info(`last sequenceNumber: ${lastSequenceNumber} was fetched from object storage`);
+
+  await runOsmdbtCommand(OSMDBT_GET_LOG, ['-m', osmdbtConfig.getLogMaxChanges]);
+  await runOsmdbtCommand(OSMDBT_CREATE_DIFF);
+
+  const newSequenceNumber = await getSequenceNumber();
+  if (lastSequenceNumber === newSequenceNumber) {
+    logger.info(`no diff was found on this job, exiting gracefully`);
+    processExitSafely(ExitCodes.SUCCESS);
+  }
+
+  logger.info(`diff was created for sequenceNumber: ${newSequenceNumber}, starting the upload`);
+
+  const [top, bottom, stateNumber] = getDiffDirPathComponents(newSequenceNumber);
+  const newDiffAndStatePaths = [STATE_FILE, DIFF_FILE_EXTENTION].map((fileExtention) => path.join(top, bottom, `${stateNumber}.${fileExtention}`));
+  newDiffAndStatePaths.forEach(async (filePath) => {
+    const localPath = path.join(osmdbtConfig.changesDir, filePath);
+    const fileStream = fs.createReadStream(localPath);
+    await putObjectWrapper(filePath, fileStream);
+  });
```

Comment:
this can be extracted to a function for better readability. it can also happen async. i shouldn't run in this kind of loop.

Reply:
done.